swarm/storage/mru/update.go

```diff
@@ -0,0 +1,320 @@
+
+package mru
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/multihash"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+// Signature is an alias for a static byte array with the size of a signature
+const signatureLength = 65
+
+type Signature [signatureLength]byte
+
+// SignedResourceUpdate contains signature information about a resource update
+type SignedResourceUpdate struct {
+	resourceUpdate // actual content that will be put on the chunk, less signature
+	signature      *Signature
+	updateAddr     storage.Address // resulting chunk address for the update
+}
+
+// Update chunk layout
+// Header:
+// 2 bytes header Length
+// 2 bytes data length
+// 4 bytes period
+// 4 bytes version
+// 32 bytes rootAddr reference
+// 32 bytes metaHash digest
+// Data:
+// data (datalength bytes)
+// Signature:
+// signature: 65 bytes (signatureLength constant)
+//
+// Minimum size is Header + 1 (minimum data length, enforced) + Signature = 142 bytes
+
+const minimumUpdateDataLength = 142
+const maxUpdateDataLength = chunkSize - minimumUpdateDataLength
+
+// Verify checks that signatures are valid and that the signer owns the resource to be updated
+func (r *SignedResourceUpdate) Verify() (err error) {
+	if len(r.data) == 0 {
+		return NewError(ErrInvalidValue, "I refuse to waste swarm space for updates with empty values, amigo (data length is 0)")
+	}
+	if r.signature == nil {
+		return NewError(ErrInvalidSignature, "Missing signature field")
+	}
+
+	digest := resourceUpdateChunkDigest(r.updateAddr, r.metaHash, r.data)
+
+	// get the address of the signer (which also checks that it's a valid signature)
+	ownerAddr, err := getAddressFromDataSig(digest, *r.signature)
+	if err != nil {
+		return err
+	}
+
+	if !bytes.Equal(r.updateAddr, r.GetUpdateAddr()) {
+		return NewError(ErrInvalidSignature, "Signature address does not match with ownerAddr")
+	}
+
+	// Check if who signed the resource update really owns the resource
+	if !verifyResourceOwnership(ownerAddr, r.metaHash, r.rootAddr) {
+		return NewError(ErrUnauthorized, fmt.Sprintf("signature is valid but signer does not own the resource: %v", err))
+	}
+
+	return nil
+}
+
+// Sign executes the signature to validate the resource
+func (r *SignedResourceUpdate) Sign(signer Signer) error {
+
+	updateAddr := r.GetUpdateAddr()
+
+	digest := resourceUpdateChunkDigest(updateAddr, r.metaHash, r.data)
+	signature, err := signer.Sign(digest)
+	if err != nil {
+		return err
+	}
+	ownerAddress, err := getAddressFromDataSig(digest, signature)
+	if err != nil {
+		return NewError(ErrInvalidSignature, "Error verifying signature")
+	}
+	if ownerAddress != signer.Address() {
+		return NewError(ErrInvalidSignature, "Signer address does not match private key")
```

Comment:
Signer address does not match owner address

Reply:
Thanks. Fixed in #704